service-app/internal/api/client_test.go

```diff
@@ -0,0 +1,176 @@
+package api
+
+import (
+	"encoding/json"
+	"encoding/xml"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/ministryofjustice/opg-scanning/config"
+	"github.com/ministryofjustice/opg-scanning/internal/httpclient"
+	"github.com/ministryofjustice/opg-scanning/internal/logger"
+	"github.com/ministryofjustice/opg-scanning/internal/types"
+)
+
+type requestCaseStub struct {
+	name        string
+	xmlPayload  string
+	expectedReq *types.ScannedCaseRequest
+	expectedErr error
+}
+
+func TestCreateStubCase(t *testing.T) {
+	docTypes := []string{"COPORD", "EPA", "EP2PG", "LP1F", "LPA002", "LP1H", "LP2"}
+
+	// Test with CaseNo
+	for _, docType := range docTypes {
+		withCaseNoPayload := fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
+	<Set xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="SET.xsd">
+		<Header CaseNo="123" Scanner="9" ScanTime="2014-09-26T12:38:53" ScannerOperator="Administrator" Schedule="02-0001112-20160909185000" FeeNumber="1234"/>
+		<Body>
+			<Document Type="%s" Encoding="UTF-8" NoPages="19"></Document>
+		</Body>
+	</Set>`, docType)
+
+		runStubCaseTest(t, withCaseNoPayload, true, docType)
+	}
+
+	// Test without CaseNo
+	for _, docType := range docTypes {
+		withoutCaseNoPayload := fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
+	<Set xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="SET.xsd">
+		<Header CaseNo="" Scanner="9" ScanTime="2014-09-26T12:38:53" ScannerOperator="Administrator" Schedule="02-0001112-20160909185000" FeeNumber="1234"/>
+		<Body>
+			<Document Type="%s" Encoding="UTF-8" NoPages="19"></Document>
+		</Body>
+	</Set>`, docType)
+
+		runStubCaseTest(t, withoutCaseNoPayload, false, docType)
+	}
+}
+
+func runStubCaseTest(t *testing.T, payload string, withCaseNo bool, docType string) {
+	tests := buildTestCases(withCaseNo, docType, payload)
+
+	for _, tt := range tests {
```

Comment:
Having this as a nested function, which calls another function containing a match statement, and then does iteration, makes these tests quite hard to read.

Maybe the payloads could be made local constants, and incorporated into `buildTestCases`, so that can just return an array, e.g.

```go
func buildTestCases() []requestCaseStub {
	return []requestCaseStub{
		{
			name:       "Order Case with CaseNo",
			xmlPayload: withCaseNoPayload,
			docType:    "COPORD",
			expectedReq: &types.ScannedCaseRequest{
				CourtReference: "123",
				CaseType:       "order",
			},
			expectedErr: nil,
		},
		{
			name:        "Invalid CaseNo for non-COPORD",
			xmlPayload:  withCaseNoPayload,
			docType:     "EPA",
			expectedReq: nil,
			expectedErr: fmt.Errorf("expected error"),
		},
		...
	}
}
```

It might lead to a bit of duplication, particularly for the LPA/EPA missing case number scenarios, but I think readability trumps that

Reply:
I get it, will refactor.